### ISSUE TITLE
Geometry Shaders

### ISSUE THREAD

**Issue by benonymity** [openglonmetal/MGL#30]
What would be needed to implement support for geometry shaders in this project?

**Comment by darkaegisagain** [comment]
I haven’t looked into it much, but I tried adding geometry shaders to the GLSL to MSL compiler and the output was blank… but I had a lot of other work to do at the time so I didn’t go much further.

If you are interested I can offer support, you need to get the build up and running in Xcode to work on it but the instructions should all be there.

Mike

**Comment by darkaegisagain** [comment]
I think the best way to do geometry shaders is to use the Mesa LLVM path which allows you to run them on the host then pass the output of the geometry shader to the GPU using a pass through vertex shader.

Since Metal only supports compute shaders (which I think is the right way to go) I didn't put much thought into geometry shaders since most of what you need (if not all) can be done with compute shaders in GLSL.

**Comment by benonymity** [comment]
I'm no expert on graphics libraries, but I've watched the progress of this project eagerly; it would be great to have a translation layer for existing OpenGL applications to run on MacOS! I noticed one of the barriers to integrating this project to mature libraries like LWJGL was a lack of geometry shaders. [LWJGL #718](https://github.com/LWJGL/lwjgl3/issues/718) Do you think that any sort of integration would ever be possible, or are there too many incompatibilities? I assume Metal has support for things like multidraw, virtual memory, advanced shaders, FBOs, etc., but I don't know how much of a headache it would be to try to get it to talk fluently to programs written for OpenGL. Great work already done though, thanks for this awesome project!

**Comment by darkaegisagain** [comment]
Well I was in the room at Apple when the discussion of implementing geometry shaders in Metal and in general geometry shaders are kind of a hack when you compare them to compute shaders. Even our hardware vendors believe that compute shaders are the way forward.

I worked on OpenCL before it was released in 2008 and one of the demos for WWDC I wrote was a full blown subdivision surface shader which handed vertices from OpenCL to OpenGL, something that you couldn't have done with geometry shaders before they became part of OpenGL.

I think supporting geometry shaders could be done, you would have to write follow my recommendation above to make it work, but I also believe the path forward for OpenGL is compute shaders and tessellation shaders not geometry shaders.
